Gaudi backend doesnt work with `HF_HUB_OFFLINE=1`

### System Info

When running TGI-gaudi with `HF_HUB_OFFLINE=1` and mounting a volume with the model files already downloaded, we get an error when launching the server.

Its due to the fact that:
1. `text_generation_server/models/causal_lm.py` tries to call `get_repo_root()` which uses `optimum/habana/checkpoint_utils.py` function, which inherently reaches out to the hub.
2. I'm guessing this occurs because the `get_repo_root()` function is being passed the model_id and [the logic there](https://github.com/huggingface/optimum-habana/blob/30022ab3413626710a9a4057dd669ca20912bbcc/optimum/habana/checkpoint_utils.py#L15) is only checking if this is a Path.

I've verified that if we overwrite the `MODEL_ID` with the absolute path to the model snapshot on disk, everything works.

Here's the full error trace:
```
2025-04-24T18:47:49.143256Z  INFO text_generation_launcher: Args {
    model_id: "meta-llama/Meta-Llama-3.1-8B-Instruct",
    revision: None,
    validation_workers: 2,
    sharded: None,
    num_shard: None,
    quantize: None,
    speculate: None,
    dtype: None,
    kv_cache_dtype: None,
    trust_remote_code: false,
    max_concurrent_requests: 64,
    max_best_of: 2,
    max_stop_sequences: 4,
    max_top_n_tokens: 5,
    max_input_tokens: Some(
        1024,
    ),
    max_input_length: None,
    max_total_tokens: Some(
        2048,
    ),
    waiting_served_ratio: 1.2,
    max_batch_prefill_tokens: Some(
        2048,
    ),
    max_batch_total_tokens: None,
    max_waiting_tokens: 7,
    max_batch_size: Some(
        32,
    ),
    cuda_graphs: None,
    hostname: "480c22f42f39",
    port: 80,
    shard_uds_path: "/tmp/text-generation-server",
    master_addr: "localhost",
    master_port: 29500,
    huggingface_hub_cache: None,
    weights_cache_override: None,
    disable_custom_kernels: false,
    cuda_memory_fraction: 1.0,
    rope_scaling: None,
    rope_factor: None,
    json_output: false,
    otlp_endpoint: None,
    otlp_service_name: "text-generation-inference.router",
    cors_allow_origin: [],
    api_key: None,
    watermark_gamma: None,
    watermark_delta: None,
    ngrok: false,
    ngrok_authtoken: None,
    ngrok_edge: None,
    tokenizer_config_path: None,
    disable_grammar_support: false,
    env: false,
    max_client_batch_size: 4,
    lora_adapters: None,
    usage_stats: On,
    payload_limit: 2000000,
    enable_prefill_logprobs: false,
    graceful_termination_timeout: 90,
}
2025-04-24T18:47:49.143354Z  INFO hf_hub: Using token file found "/data/token"    
2025-04-24T18:47:50.292338Z  WARN text_generation_launcher::gpu: Cannot determine GPU compute capability: AssertionError: Torch not compiled with CUDA enabled
2025-04-24T18:47:50.292361Z  INFO text_generation_launcher: Using attention default - Prefix caching 0
2025-04-24T18:47:50.292376Z  INFO text_generation_launcher: Using default cuda graphs [1, 2, 4, 8, 16, 32]
2025-04-24T18:47:50.292489Z  INFO download: text_generation_launcher: Starting check and download process for meta-llama/Meta-Llama-3.1-8B-Instruct
2025-04-24T18:47:54.310795Z  INFO text_generation_launcher: Files are already present on the host. Skipping download.
2025-04-24T18:47:54.912484Z  INFO download: text_generation_launcher: Successfully downloaded weights for meta-llama/Meta-Llama-3.1-8B-Instruct
2025-04-24T18:47:54.913182Z  INFO shard-manager: text_generation_launcher: Starting shard rank=0
2025-04-24T18:47:58.312814Z  INFO text_generation_launcher: Using prefix caching = False
2025-04-24T18:47:58.312861Z  INFO text_generation_launcher: Using Attention = default
2025-04-24T18:48:00.174248Z  WARN text_generation_launcher: FBGEMM fp8 kernels are not installed.
2025-04-24T18:48:00.180375Z  INFO text_generation_launcher: quantize=None
2025-04-24T18:48:00.180457Z  INFO text_generation_launcher: CLI SHARDED = False DTYPE = bfloat16
2025-04-24T18:48:00.392817Z  INFO text_generation_launcher: Server:server_inner: sharded =False
2025-04-24T18:48:00.392973Z  INFO text_generation_launcher: Server:server_inner: data type = bfloat16, local_url = unix:///tmp/text-generation-server-0
2025-04-24T18:48:00.828931Z ERROR text_generation_launcher: Error when initializing model
Traceback (most recent call last):
  File "/usr/local/bin/text-generation-server", line 8, in <module>
    sys.exit(app())
  File "/usr/local/lib/python3.10/dist-packages/typer/main.py", line 311, in __call__
    return get_command(self)(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/typer/core.py", line 778, in main
    return _main(
  File "/usr/local/lib/python3.10/dist-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.10/dist-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/cli.py", line 173, in serve
    server.serve(
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/server.py", line 279, in serve
    asyncio.run(
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 1909, in _run_once
    handle._run()
  File "/usr/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
> File "/usr/local/lib/python3.10/dist-packages/text_generation_server/server.py", line 233, in serve_inner
    model = get_model_with_lora_adapters(
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/models/__init__.py", line 242, in get_model_with_lora_adapters
    model = get_model(
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/models/__init__.py", line 215, in get_model
    return CausalLM(
  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/models/causal_lm.py", line 718, in __init__
    get_repo_root(model_id)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/checkpoint_utils.py", line 26, in get_repo_root
    ".safetensors" in filename for filename in list_repo_files(model_name_or_path, token=token)
  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/hf_api.py", line 2935, in list_repo_files
    return [
  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/hf_api.py", line 2935, in <listcomp>
    return [
  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/hf_api.py", line 3072, in list_repo_tree
    for path_info in paginate(path=tree_url, headers=headers, params={"recursive": recursive, "expand": expand}):
  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_pagination.py", line 36, in paginate
    r = session.get(path, params=params, headers=headers)
  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 602, in get
    return self.request("GET", url, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_http.py", line 104, in send
    raise OfflineModeIsEnabled(
huggingface_hub.errors.OfflineModeIsEnabled: Cannot reach https://huggingface.co/api/models/meta-llama/Meta-Llama-3.1-8B-Instruct/tree/main?recursive=True&expand=False: offline mode is enabled. To disable it, please unset the `HF_HUB_OFFLINE` environment variable.
2025-04-24T18:48:01.733504Z ERROR shard-manager: text_generation_launcher: Shard complete standard error output:

2025-04-24 18:47:56.238 | INFO     | text_generation_server.utils.import_utils:<module>:75 - Detected system cpu
/usr/local/lib/python3.10/dist-packages/text_generation_server/utils/sgmv.py:18: UserWarning: Could not import SGMV kernel from Punica, falling back to loop.
  warnings.warn("Could not import SGMV kernel from Punica, falling back to loop.")
/usr/local/lib/python3.10/dist-packages/transformers/deepspeed.py:24: FutureWarning: transformers.deepspeed module is deprecated and will be removed in a future version. Please import deepspeed modules directly from transformers.integrations
  warnings.warn(
Traceback (most recent call last):

  File "/usr/local/bin/text-generation-server", line 8, in <module>
    sys.exit(app())

  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/cli.py", line 173, in serve
    server.serve(

  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/server.py", line 279, in serve
    asyncio.run(

  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)

  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()

  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/server.py", line 233, in serve_inner
    model = get_model_with_lora_adapters(

  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/models/__init__.py", line 242, in get_model_with_lora_adapters
    model = get_model(

  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/models/__init__.py", line 215, in get_model
    return CausalLM(

  File "/usr/local/lib/python3.10/dist-packages/text_generation_server/models/causal_lm.py", line 718, in __init__
    get_repo_root(model_id)

  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/checkpoint_utils.py", line 26, in get_repo_root
    ".safetensors" in filename for filename in list_repo_files(model_name_or_path, token=token)

  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)

  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/hf_api.py", line 2935, in list_repo_files
    return [

  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/hf_api.py", line 2935, in <listcomp>
    return [

  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/hf_api.py", line 3072, in list_repo_tree
    for path_info in paginate(path=tree_url, headers=headers, params={"recursive": recursive, "expand": expand}):

  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_pagination.py", line 36, in paginate
    r = session.get(path, params=params, headers=headers)

  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 602, in get
    return self.request("GET", url, **kwargs)

  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)

  File "/usr/local/lib/python3.10/dist-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)

  File "/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_http.py", line 104, in send
    raise OfflineModeIsEnabled(

huggingface_hub.errors.OfflineModeIsEnabled: Cannot reach https://huggingface.co/api/models/meta-llama/Meta-Llama-3.1-8B-Instruct/tree/main?recursive=True&expand=False: offline mode is enabled. To disable it, please unset the `HF_HUB_OFFLINE` environment variable.
 rank=0
Error: ShardCannotStart
2025-04-24T18:48:01.819511Z ERROR text_generation_launcher: Shard 0 failed to start
2025-04-24T18:48:01.819543Z  INFO text_generation_launcher: Shutting down shards
```

### Information

- [x] Docker
- [ ] The CLI directly

### Tasks

- [x] An officially supported command
- [ ] My own modifications

### Reproduction

Run TGI with `HF_HUB_OFFLINE=1` env var:

```
docker run -p 8080:80 \
   --runtime=habana \
   --cap-add=sys_nice \
   --ipc=host \
   -v $volume:/data \
   -e HF_HUB_OFFLINE=1 \
   -e MAX_TOTAL_TOKENS=2048 \
   -e PREFILL_BATCH_BUCKET_SIZE=2 \
   -e BATCH_BUCKET_SIZE=32 \
   -e PAD_SEQUENCE_TO_MULTIPLE_OF=256 \
   ghcr.io/huggingface/text-generation-inference:3.2.3-gaudi \
   --model-id $model \
   --max-input-tokens 1024 \
   --max-total-tokens 2048 \
   --max-batch-prefill-tokens 2048 \
   --max-batch-size 32 \
   --max-waiting-tokens 7 \
   --waiting-served-ratio 1.2 \
   --max-concurrent-requests 64
```

### Expected behavior

TGI-gaudi should be able to gracefully handle offline mode when the `model_id` is provided, findable, and mounted to the running container. This would reach parity with how other TGI backends work.

--- a/backends/gaudi/Makefile
+++ b/backends/gaudi/Makefile
@@ -8,7 +8,7 @@ PYTORCH_VERSION := 2.6.0
 .PHONY:	image run-local-dev-container install-dependencies install-server install-router install-launcher local-dev-install
 
 image:
-	docker build -t tgi-gaudi -f ${root_dir}/Dockerfile_gaudi ${root_dir} --build-arg HABANA_VERSION=$(HABANA_VERSION) --build-arg PYTORCH_VERSION=$(PYTORCH_VERSION)
+	docker build --ulimit nofile=4096 -t tgi-gaudi -f ${root_dir}/Dockerfile_gaudi ${root_dir} --build-arg HABANA_VERSION=$(HABANA_VERSION) --build-arg PYTORCH_VERSION=$(PYTORCH_VERSION)
 
 run-local-dev-container:
 		docker run -it \

--- a/backends/gaudi/server/text_generation_server/models/causal_lm.py
+++ b/backends/gaudi/server/text_generation_server/models/causal_lm.py
@@ -4,6 +4,7 @@ import bisect
 from dataclasses import dataclass
 from functools import wraps
 import itertools
+import json
 import math
 import os
 import tempfile
@@ -17,15 +18,12 @@ from loguru import logger
 from opentelemetry import trace
 
 import text_generation_server.habana_quantization_env as hq_env
+from text_generation_server.utils import weight_files
 import habana_frameworks.torch as htorch
 from optimum.habana.utils import HabanaProfile
 from optimum.habana.transformers.generation import MODELS_OPTIMIZED_WITH_STATIC_SHAPES
 from text_generation_server.utils.chunks import concat_text_chunks
-from optimum.habana.checkpoint_utils import (
-    get_repo_root,
-    model_on_meta,
-    write_checkpoints_json,
-)
+from optimum.habana.checkpoint_utils import model_on_meta
 from transformers import (
     AutoTokenizer,
     AutoModelForCausalLM,
@@ -708,6 +706,9 @@ class CausalLM(Model):
         if hq_env.is_quantization_enabled:
             htorch.core.hpu_set_env()
 
+        # Get weight files
+        weight_files(model_id, revision=revision, extension=".safetensors")
+
         if world_size > 1:
             os.environ.setdefault(
                 "DEEPSPEED_USE_HABANA_FRAMEWORKS_DETERMINISTIC_API", "1"
@@ -715,8 +716,6 @@ class CausalLM(Model):
             model = self.get_deepspeed_model(model_id, dtype, revision)
             model = hq_env.prepare_model_for_quantization(model)
         else:
-            get_repo_root(model_id)
-
             # Check support for rope scaling
             model_kwargs = {}
             config = AutoConfig.from_pretrained(model_id)
@@ -868,7 +867,6 @@ class CausalLM(Model):
             with deepspeed.OnDevice(dtype=dtype, device="meta"):
                 model = AutoModelForCausalLM.from_config(config, torch_dtype=dtype)
         else:
-            get_repo_root(model_id, local_rank=os.getenv("LOCAL_RANK"))
             # TODO: revisit placement on CPU when auto-injection is possible
             with deepspeed.OnDevice(dtype=dtype, device="cpu"):
                 model = AutoModelForCausalLM.from_pretrained(
@@ -884,7 +882,11 @@ class CausalLM(Model):
         if load_to_meta:
             # model loaded to meta is managed differently
             checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="+w")
-            write_checkpoints_json(model_id, local_rank, checkpoints_json)
+            checkpoint_files = [str(f) for f in weight_files(model_id, revision=revision, extension=".safetensors")]
+            data = {"type": "ds_model", "checkpoints": checkpoint_files, "version": 1.0}
+            json.dump(data, checkpoints_json)
+            checkpoints_json.flush()
+
             ds_inference_kwargs["checkpoint"] = checkpoints_json.name
         model = deepspeed.init_inference(model, **ds_inference_kwargs)
 

--- a/backends/gaudi/server/text_generation_server/models/causal_lm.py
+++ b/backends/gaudi/server/text_generation_server/models/causal_lm.py
@@ -882,7 +882,12 @@ class CausalLM(Model):
         if load_to_meta:
             # model loaded to meta is managed differently
             checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="+w")
-            checkpoint_files = [str(f) for f in weight_files(model_id, revision=revision, extension=".safetensors")]
+            checkpoint_files = [
+                str(f)
+                for f in weight_files(
+                    model_id, revision=revision, extension=".safetensors"
+                )
+            ]
             data = {"type": "ds_model", "checkpoints": checkpoint_files, "version": 1.0}
             json.dump(data, checkpoints_json)
             checkpoints_json.flush()

--- a/backends/gaudi/server/text_generation_server/models/vlm_causal_lm.py
+++ b/backends/gaudi/server/text_generation_server/models/vlm_causal_lm.py
@@ -1,3 +1,4 @@
+import json
 import re
 import torch
 import os
@@ -12,6 +13,7 @@ import tempfile
 import copy
 from text_generation_server.models import Model
 from transformers import PreTrainedTokenizerBase
+from text_generation_server.utils import weight_files
 from text_generation_server.utils.tokens import batch_top_tokens
 from text_generation_server.pb import generate_pb2
 from text_generation_server.models.causal_lm import (
@@ -43,11 +45,7 @@ from transformers import (
     AutoTokenizer,
     AutoConfig,
 )
-from optimum.habana.checkpoint_utils import (
-    get_repo_root,
-    model_on_meta,
-    write_checkpoints_json,
-)
+from optimum.habana.checkpoint_utils import model_on_meta
 
 from text_generation_server.utils.speculate import get_speculate
 from text_generation_server.models.types import (
@@ -840,6 +838,9 @@ class VlmCausalLM(Model):
         if hq_env.is_quantization_enabled:
             htorch.core.hpu_set_env()
 
+        # Get weight files
+        weight_files(model_id, revision=revision, extension=".safetensors")
+
         if world_size > 1:
             os.environ.setdefault(
                 "DEEPSPEED_USE_HABANA_FRAMEWORKS_DETERMINISTIC_API", "1"
@@ -847,8 +848,6 @@ class VlmCausalLM(Model):
             model = self.get_deepspeed_model(model_class, model_id, dtype, revision)
             model = hq_env.prepare_model_for_quantization(model)
         else:
-            get_repo_root(model_id)
-
             # Check support for rope scaling
             model_kwargs = {}
             config = AutoConfig.from_pretrained(model_id)
@@ -1000,7 +999,6 @@ class VlmCausalLM(Model):
             with deepspeed.OnDevice(dtype=dtype, device="meta"):
                 model = model_class.from_config(config, torch_dtype=dtype)
         else:
-            get_repo_root(model_id, local_rank=os.getenv("LOCAL_RANK"))
             # TODO: revisit placement on CPU when auto-injection is possible
             with deepspeed.OnDevice(dtype=dtype, device="cpu"):
                 model = model_class.from_pretrained(
@@ -1019,7 +1017,10 @@ class VlmCausalLM(Model):
         if load_to_meta:
             # model loaded to meta is managed differently
             checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="+w")
-            write_checkpoints_json(model_id, local_rank, checkpoints_json)
+            checkpoint_files = [str(f) for f in weight_files(model_id, revision=revision, extension=".safetensors")]
+            data = {"type": "ds_model", "checkpoints": checkpoint_files, "version": 1.0}
+            json.dump(data, checkpoints_json)
+            checkpoints_json.flush()
             ds_inference_kwargs["checkpoint"] = checkpoints_json.name
         model = deepspeed.init_inference(model, **ds_inference_kwargs)
 

--- a/backends/gaudi/server/text_generation_server/models/vlm_causal_lm.py
+++ b/backends/gaudi/server/text_generation_server/models/vlm_causal_lm.py
@@ -1017,7 +1017,12 @@ class VlmCausalLM(Model):
         if load_to_meta:
             # model loaded to meta is managed differently
             checkpoints_json = tempfile.NamedTemporaryFile(suffix=".json", mode="+w")
-            checkpoint_files = [str(f) for f in weight_files(model_id, revision=revision, extension=".safetensors")]
+            checkpoint_files = [
+                str(f)
+                for f in weight_files(
+                    model_id, revision=revision, extension=".safetensors"
+                )
+            ]
             data = {"type": "ds_model", "checkpoints": checkpoint_files, "version": 1.0}
             json.dump(data, checkpoints_json)
             checkpoints_json.flush()

--- a/router/src/server.rs
+++ b/router/src/server.rs
@@ -1577,7 +1577,7 @@ pub async fn run(
             let cache = std::env::var("HUGGINGFACE_HUB_CACHE")
                 .map_err(|_| ())
                 .map(|cache_dir| Cache::new(cache_dir.into()))
-                .unwrap_or_else(|_| Cache::default());
+                .unwrap_or_else(|_| Cache::from_env());
             tracing::warn!("Offline mode active using cache defaults");
             Type::Cache(cache)
         } else {

